File: apps/common/src/main/java/ro/mig/common/Artefacts.java
package ro.mig.common;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.InputStream;
import java.io.UncheckedIOException;
import java.io.IOException;
import java.util.Map;

/**
 * The five artefact types from the architecture diagram, and the object naming
 * convention shared by every component that reads or writes them.
 *
 * <pre>
 *   .DAT  flat file, source data (split files if needed)
 *   .CHS  checksums for reconciliation
 *   .ERR  errors
 *   .RPT  report files (#records etc)
 *   .FLG  flag file — the semaphore that all files were generated
 * </pre>
 *
 * <p>
 * Both outer lanes produce the same five types: extraction on the way in, load
 * on the way out. The {@code lane} prefix keeps them apart in the bucket.
 *
 * <p>
 * <b>The convention is not written here.</b> It is loaded from
 * {@code contracts/artefacts.json}, the same file
 * {@code pipelines/common/artefacts.py} reads — Maven copies it onto this
 * module's classpath at build time. Until that manifest existed, the convention
 * was declared twice, once per language, and kept in step by a comment in each;
 * a rename on the Python side stayed invisible here until a run failed at
 * integration.
 */
public final class Artefacts {

    private static final JsonNode MANIFEST = load();

    public static final String LANE_EXTRACTION = lane("extraction");
    public static final String LANE_LOAD = lane("load");

    private Artefacts() {
    }

    private static JsonNode load() {
        try (InputStream in = Artefacts.class.getResourceAsStream("/artefacts.json")) {
            if (in == null) {
                // A hardcoded fallback would silently reintroduce the drift this class
                // exists to prevent, so an absent manifest is fatal rather than papered
                // over: it means the jar was built without contracts/artefacts.json.
                throw new IllegalStateException(
                        "artefacts.json is not on the classpath — the jar was built without "
                                + "contracts/artefacts.json; check apps/common/pom.xml <resources>");
            }
            return new ObjectMapper().readTree(in);
        } catch (IOException e) {
            throw new UncheckedIOException("cannot read the artefact manifest", e);
        }
    }

    private static String template(String key) {
        JsonNode node = MANIFEST.path("artefacts").path(key);
        if (node.isMissingNode()) {
            throw new IllegalStateException("artefact manifest declares no '" + key + "'");
        }
        return node.asText();
    }

    private static String lane(String key) {
        JsonNode node = MANIFEST.path("lanes").path(key);
        if (node.isMissingNode()) {
            throw new IllegalStateException("artefact manifest declares no lane '" + key + "'");
        }
        return node.asText();
    }

    /** Substitutes {@code {record}} / {@code {sequence}} / {@code {lane}} / {@code {run_id}}. */
    private static String fill(String template, Map<String, String> values) {
        String out = template;
        for (Map.Entry<String, String> e : values.entrySet()) {
            out = out.replace("{" + e.getKey() + "}", e.getValue());
        }
        return out;
    }

    /** e.g. {@code extraction/initial-1234-ab12/} */
    public static String prefix(String lane, String runId) {
        return fill(MANIFEST.path("prefix").asText(), Map.of("lane", lane, "run_id", runId));
    }

    /**
     * Split data files are sequenced, so a large extract can be produced in parts.
     * The pad width comes from the manifest, so both languages zero-pad alike.
     */
    public static String dat(String record, int sequence) {
        int width = MANIFEST.path("dat_sequence_width").asInt();
        return fill(template("dat"), Map.of(
                "record", record,
                "sequence", String.format("%0" + width + "d", sequence)));
    }

    public static String chs(String record) {
        return fill(template("chs"), Map.of("record", record));
    }

    public static String err(String record) {
        return fill(template("err"), Map.of("record", record));
    }

    public static String rpt(String record) {
        return fill(template("rpt"), Map.of("record", record));
    }

    /**
     * The semaphore. Written <em>last</em> and outside the encrypted bundle, so a
     * consumer can detect "the extract is complete" without holding a decryption key.
     */
    public static String flg(String record) {
        return fill(template("flg"), Map.of("record", record));
    }

    /**
     * Archived, compressed and PGP-encrypted transport unit holding DAT/CHS/ERR/RPT.
     */
    public static String bundle(String record) {
        return fill(template("bundle"), Map.of("record", record));
    }

    public static boolean isDat(String name) {
        return name.endsWith(".DAT");
    }

    /**
     * Physical BigQuery column name for a logical field — {@code table} is
     * {@code src} or {@code target}. recon-service reads both sets by name, which is
     * why the mapping belongs in the manifest and not in either language's source.
     */
    public static String column(String table, String logical) {
        JsonNode node = MANIFEST.path("columns").path(table).path(logical);
        if (node.isMissingNode()) {
            throw new IllegalStateException(
                    "artefact manifest declares no column '" + table + "." + logical + "'");
        }
        return node.asText();
    }
}

File: pipelines/common/artefacts.py
"""The artefact naming convention and shared BigQuery column names — loaded, not restated.

Both this module and `apps/common/.../Artefacts.java` read `contracts/artefacts.json`.
Before that file existed the convention lived twice, once per language, kept in step by a
comment in each — so a rename in Python stayed invisible to Java until a run failed at
integration. The manifest makes the agreement a file both sides load rather than a promise
both sides make.

Resolution order for the contracts directory: `MIG_CONTRACTS_DIR` if set (the Dataflow
images place it elsewhere), otherwise `contracts/` next to the repo root. The Java side does
not need a path at all — Maven copies the same file onto its classpath.
"""

from __future__ import annotations

import json
import os
from functools import lru_cache
from pathlib import Path
from typing import Any

_REPO_ROOT = Path(__file__).resolve().parents[2]
MANIFEST_NAME = "artefacts.json"


def manifest_path() -> Path:
    override = os.environ.get("MIG_CONTRACTS_DIR", "").strip()
    base = Path(override) if override else _REPO_ROOT / "contracts"
    return base / MANIFEST_NAME


@lru_cache(maxsize=1)
def manifest() -> dict[str, Any]:
    """The parsed manifest. Cached — it is read once per process and never changes."""
    path = manifest_path()
    if not path.is_file():
        # Failing loudly here beats falling back to a hardcoded convention: a silent
        # fallback is exactly the drift this file exists to prevent.
        raise FileNotFoundError(
            f"artefact manifest not found at {path} — set MIG_CONTRACTS_DIR if the "
            f"contracts directory is elsewhere"
        )
    return json.loads(path.read_text(encoding="utf-8"))


def prefix(lane: str, run_id: str) -> str:
    """e.g. `extraction/initial-20260818-081725/`."""
    lanes = manifest()["lanes"]
    if lane not in lanes:
        raise ValueError(f"unknown lane {lane!r} — the manifest declares {sorted(lanes)}")
    return manifest()["prefix"].format(lane=lanes[lane], run_id=run_id)


def dat(record: str, sequence: int) -> str:
    """Sequenced data file — a large extract can be produced in parts."""
    m = manifest()
    width = m["dat_sequence_width"]
    return m["artefacts"]["dat"].format(record=record, sequence=f"{sequence:0{width}d}")


def artefact_names(record: str) -> dict[str, str]:
    """Every non-sequenced artefact for one record, keyed by type."""
    return {
        key: template.format(record=record)
        for key, template in manifest()["artefacts"].items()
        if key != "dat"
    }


def columns(table: str) -> dict[str, str]:
    """Logical name → physical BigQuery column, for `src` or `target`."""
    cols = manifest()["columns"]
    if table not in cols:
        raise ValueError(f"unknown table {table!r} — the manifest declares {sorted(cols)}")
    return dict(cols[table])
